### PR TITLE
#fixed Some crashes

### PR DESCRIPTION
**Pull request by jamesstout**
## Changes:
- see commit messages

## Closes following issues:
- FB: 6d757836480c5d7ed457435ddf60eddb
- FB: 4cd426bed28002865565f7e7e8b4eb53
- FB: 246e9c940b6aac10e827d546de79da8d
- FB: 921a2f25e4ed066aa0289c85eb297dec
- FB: 5bd264a2ea5ccd2291f1d3911b41e1e7
- FB: 76f4b3b6acb0db0542ca2d878fc5fa85
- FB: 22ff32ad5818cadceb304673750a86b9
- FB: 36eed87ab07afdcfab8efe1a42b07719

## Tested:
- Processors:
  - [x] Intel
  - [ ] Apple Silicon
- macOS Versions:
  - [ ] 10.12.x (Sierra)
  - [ ] 10.13.x (High Sierra)
  - [ ] 10.14.x (Mojave)
  - [x] 10.15.x (Catalina)
  - [ ] 11.x (Big Sur)
- Xcode Version: 12.3 (12C33)
- Localizations:
  - [ ] English
  - [ ] Spanish
  - [ ] Other (please specify)

### CODE REVIEW

File: Source/Other/Parsing/SPSQLParser.m [hunk 1]
```diff
@@ -1009,6 +1010,10 @@ - (NSUInteger) length {
 	return [string length];
 }
 - (unichar) characterAtIndex:(NSUInteger)anIndex {
+    if(anIndex >= string.length){
+        CLS_LOG(@"ERROR: anIndex: [%lu] >= string.len: [%lu]. string = %@", (unsigned long)anIndex, (unsigned long)string.length, safeString(string));
+        return (unichar)0x00; // FIXME: what should this return?
```
Comment (Kaspik) [inline Source/Other/Parsing/SPSQLParser.m:1015]: no ieda

File: Source/Controllers/Preferences/Panes/SPFilePreferencePane.m [hunk 2]
```diff
@@ -480,37 +480,45 @@ - (void)chooseFileWithOptions:(PanelOptions*)options
...
-                    SPLog(@"self->selectedRows: %@", self->selectedRows);
-
-                    // increment for the next file
-                    options.index++;
-                }
+            // check it's really a URL
+            if(![self->_currentFilePanel.URL isKindOfClass:[NSURL class]]){
+                SPLog(@"self->keySelectionPanel.URL is not a URL: %@", self->_currentFilePanel.URL.class);
+                CLS_LOG(@"self->keySelectionPanel.URL is not a URL: %@", self->_currentFilePanel.URL.class);
+                // JCS - should we stop here?
```
Comment (Kaspik) [inline Source/Controllers/Preferences/Panes/SPFilePreferencePane.m:487]: Probably?

**Review by Kaspik — COMMENTED** [review]
Will leave the review up to others

File: Source/Other/Parsing/SPSQLParser.m (hunk 1 above)
Comment (jamesstout) [reply Source/Other/Parsing/SPSQLParser.m:1015]: I think null char is ok, we'll catch errors.

File: Source/Controllers/Preferences/Panes/SPFilePreferencePane.m (hunk 2 above)
Comment (jamesstout) [reply Source/Controllers/Preferences/Panes/SPFilePreferencePane.m:487]: reviewing at the moment...

Comment (jamesstout) [reply Source/Controllers/Preferences/Panes/SPFilePreferencePane.m:487]: Ah, just the act of selecting the file in the NSOpenPanel calls `startAccessingSecurityScopedResource`

I swizzled `startAccessingSecurityScopedResource`. This is output as you choose the file in the export window:

```log
2021-01-15 19:18:43.301411+0800 Sequel Ace[54819:3312141] -[NSURL(SPURLAdditions)
 SA_startAccessingSecurityScopedResource]:36: JIMMY startAccessingSecurityScopedResource called:
 file:///Users/james/untitled%20folder%202/untitled%20folder/untitled%20folder/
```

So the only downside of this failing is that we won't have a bookmark, and therefore no access on the next app start. I don't think we should stop.

Comment (jamesstout) [reply Source/Controllers/Preferences/Panes/SPFilePreferencePane.m:487]: Oh, hold on I had two places where I put "JCS - should we stop here?" ... 

1. Not a valid URL
2. Couldn't create a bookmark.

I was talking about 2 above...

Comment (jamesstout) [reply Source/Controllers/Preferences/Panes/SPFilePreferencePane.m:487]: But still .. .if it's not a URL, we don't try to create a bookmark ... but then how does the exporter (or whatever) read/write to the file? Need an example of an invalid URL....

Comment (jamesstout) [reply Source/Controllers/Preferences/Panes/SPFilePreferencePane.m:487]: so the logging I added in this PR should give us some examples.

**Review by Jason-Morcos — APPROVED** [review]

**Pull request merged**